src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs

```diff
@@ -0,0 +1,177 @@
+        {
+            _serviceMappingMessages.AddOrUpdate(
+                serviceMappingMessage.InstanceId,
+                new List<string>() { serviceMappingMessage.InvocationId },
+                (key, valueList) => { valueList.Add(serviceMappingMessage.InvocationId); return valueList; });
+        }
+
+        public void CleanupInvocations(string instanceId)
+        {
+            if (_serviceMappingMessages.TryRemove(instanceId, out var invocationIds))
+            {
+                foreach (var invocationId in invocationIds)
+                {
+                    if (_pendingInvocations.TryRemove(invocationId, out var item))
+                    {
+                        var message = new CompletionMessage(invocationId, $"Connection '{item.ConnectionId}' is disconnected.", null, false);
+                        item.Complete(item.Tcs, message);
+                    }
+                }
+            }
+        }
+
+        public bool TryCompleteResult(string connectionId, CompletionMessage message)
+        {
+            if (_pendingInvocations.TryGetValue(message.InvocationId!, out var item))
+            {
+                if (item.ConnectionId != connectionId)
+                {
+                    throw new InvalidOperationException($"Connection ID '{connectionId}' is not valid for invocation ID '{message.InvocationId}'.");
+                }
+
+                // if false the connection disconnected right after the above TryGetValue
+                // or someone else completed the invocation (likely a bad client)
+                // we'll ignore both cases
+                if (_pendingInvocations.TryRemove(message.InvocationId!, out _))
+                {
+                    item.Complete(item.Tcs, message);
+                    return true;
+                }
+                return false;
+            }
+            else
+            {
+                // connection was disconnected or someone else completed the invocation
+                return false;
+            }
+        }
+
+        public bool TryCompleteResult(string connectionId, ClientCompletionMessage message)
+        {
+            var protocol = _hubProtocolResolver.GetProtocol(message.Protocol, null);
+            if (protocol == null)
+            {
+                throw new InvalidOperationException($"Not supported protocol {message.Protocol} by server");
+            }
+
+            var payload = message.Payload;
+            if (protocol.TryParseMessage(ref payload, this, out var hubMessage))
+            {
+                if (hubMessage is CompletionMessage completionMessage)
+                {
+                    return TryCompleteResult(connectionId, completionMessage);
+                }
+                else
+                {
+                    throw new InvalidOperationException("The payload of ClientCompletionMessage cannot be parsed into CompletionMessage correctly.");
```

Comment:
also add hubMessage.GetType into the exception

Reply:
done